pkg/operator/baremetal_config.go

```diff
@@ -34,6 +34,15 @@ type BaremetalProvisioningConfig struct {
 	ProvisioningOSDownloadURL string
 }
 
+func isBaremetalProvisioningConfigAvailable(dc dynamic.Interface, configName string) bool {
+	provisioningClient := dc.Resource(provisioningGVR)
+	_, err := provisioningClient.Get(context.Background(), configName, metav1.GetOptions{})
+	if err != nil {
```

Comment:
In a baremetal cluster, will the provisioning CRDs always be deployed, or will they only be deployed in the case that Metal3 is desired? Just wondering if we should distinguish the behaviour of, for example, a failed connection vs the GVR missing from the cluster vs the configmap not existing

Reply:
@JoelSpeed Currently, provisioning CR is always deployed when the platform is baremetal. With the need to support this new BareMetal installation type where the platform type is Baremetal but there is no need for the metal3 pod to manage the cluster, we would not need the provisioning CR.
With this change, within the MAO, we will not be able to distinguish between the cases where there was some failure deploying the provisioning CR or if we do not need the CR at all. And, that is OK I think. 
I am not sure I understand what you mean by "failed connection vs the GVR missing". Regarding the config map, since we are blocking metal3 deployment from happening without the provisioning config, the only instance where the configmap would be the case when the provisioning CR exists but is missing some values.

Reply:
>With this change, within the MAO, we will not be able to distinguish between the cases where there was some failure deploying the provisioning CR or if we do not need the CR at all. And, that is OK I think

Why is that ok? here https://github.com/openshift/machine-api-operator/pull/560/files#diff-fa45321336db7ad1cedc28bf643a4f97R60 you are using this function to decide if the pod needs to be deployed. But the function as it is, could be lying to you. Unless I'm missing anything we need to check `if err != nil  and errIsNotFound` here

Reply:
When there are intermittent failures reading the provisioning CR, then the call to isBaremetalProvisioningConfigAvailable() will succeed in a subsequent call to sync(). And that would result in the metal3 pod being deployed. 
If the provisioning CR is really not present, then in every sync() call, the decision would be made not to deploy the metal3 pod.
That is what I tried to convey in my previous comment.

Reply:
I see that it accidentally would work, that is brittle. You should account for the error semantics that are relevant here (notFound vs other failures) and let the callers react appropriately.

Also I see there's also already a func `getBaremetalProvisioningConfig` which returns an error from `provisioningClient.Get(context.Background(), configName, metav1.GetOptions{})` so I see no need for a new func, you can reuse that and react to the notFound error.

Also this code as it is is completely dropping the ability to fall back to the configMap if the ProvisioningConfig CRD does not exists so can we  please make sure to remove all that dead code?

Reply:
I added isBaremetalProvisioningConfigAvailable() as a helper function because the assisted-install team wanted a silent check and getBaremetalProvisioningConfig() is not silent when it is not able to read provisioning config. Based on previous comment, I am removing the helper function and updating getBaremetalProvisioningConfig() to not report anything when it is not able to read the provisioning CR. More code reuse for sure with slight loss of debuggability.

Also, removing any code trying to read config from a configmap as suggested above since we would not hit that scenario at all with this change.